### PR TITLE
Fix contained LEA handling

### LINKED ISSUE #25039: Specific AVX2 intrinsic inlining generates faulty x64 code
I have a short sequence of code that seems to generate faulty x64 code (and crashes the program).

I'm using the latest SDK bits as of this time as suggested by @tannergooding, `dotnet --info` says:
```bash
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview7-012287
 Commit:    f202b59402

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  19.04
 OS Platform: Linux
 RID:         ubuntu.19.04-x64
 Base Path:   /home/dmg/dotnet/sdk/3.0.100-preview7-012287/

Host (useful for support):
  Version: 3.0.0-preview7-27806-02
  Commit:  90a101062a

.NET Core SDKs installed:
  2.1.603 [/home/dmg/dotnet/sdk]
  2.1.700 [/home/dmg/dotnet/sdk]
  3.0.100-preview5-011568 [/home/dmg/dotnet/sdk]
  3.0.100-preview7-012287 [/home/dmg/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.11 [/home/dmg/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.11 [/home/dmg/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview5-19227-01 [/home/dmg/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0-preview7.19306.7 [/home/dmg/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.10 [/home/dmg/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [/home/dmg/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview5-27626-15 [/home/dmg/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0-preview7-27806-02 [/home/dmg/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

I have a [complete-repro project](https://github.com/damageboy/bugbugbug-coreclr-25039) setup for this issue.

The faulty code (as seen from lldb disassmbly is):

```asm
    0x7f91fc3d149a: shl    esi, 0x3
    0x7f91fc3d149d: movsxd rsi, esi
    0x7f91fc3d14a0: movabs rsi, 0x7f927405ece8
->  0x7f91fc3d14aa: vpmovzxbd ymm1, qword ptr [rsi + rsi] ; ymm1 = mem[0],zero,zero,zero,mem[1],zero,zero,zero,mem[2],zero,zero,zero,mem[3],zero,zero,zero,mem[4],zero,zero,zero,mem[5],zero,zero,zero,mem[6],zero,zero,zero,mem[7],zero,zero,zero
```

In this fragment, `esi` is initially the index into the permutation table I'm trying to load/convert, and is left shifted by 3 (as each entry is 8 bytes), but is then overwritten with the base address for the permutation table.

At this stage, the `vpmovzxbd [rsi + rsi]` is obviously totally bonkers and thankfully generated s segfault...

I *think* that the bug **might** be related to the base-address of the permutation array being a `ReadOnlySpan<byte>` embedded inside the executable, thanks to C# 7.3, as it seems the JIT is trying to correctly optimize it as a `constant` address during the compilation, but I could also be completely wrong about that last part.

Perviously @tannergooding opened up [#25008](https://github.com/dotnet/coreclr/issues/25008), which is somewhat related to this issue as it related to how the offset calculation is expressed in the generated code.

### PR DESCRIPTION
**Pull request by CarolEidt**
This adds an LEA case to both `LinearScan::BuildOperandUses` and `CodeGen::genConsumeRegs`.

Fix #25039

### CODE REVIEW

**Comment by CarolEidt** [comment]
@mikedn - I went ahead and looked into this.
While it is more efficient to infer where one is expecting an address, and should call `LinearScan::BuildAddrUses` and then `CodeGen::genConsumeAddress`, it is probably not worth the risk of requiring them to be in lock-step with `Lowering`. Ultimately we need a better mechanism, or should perhaps just eliminate the address-specific methods.

**Comment by mikedn** [comment]
> While it is more efficient to infer where one is expecting an address, and should call LinearScan::BuildAddrUses and then CodeGen::genConsumeAddress, it is probably not worth the risk of requiring them to be in lock-step with Lowering. Ultimately we need a better mechanism, or should perhaps just eliminate the address-specific methods.

Yes, my thought as well. I'm all for efficiency but not at the cost of making the code less reliable/difficult to maintain.

I was trying to clean up more of it in https://github.com/mikedn/coreclr/commit/87f0ef2263d26569f5e932e5ad1ca3af51870128 But of course, it's too late to do something like that for .NET 3.0. I may revisit this once I finish eliminating `GT_LIST`, when we get rid of all the `GT_LIST` "parsing" code in SIMD/HWIntrinsic code.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib ping
I've retried the arm leg without success, so my inclination is to go ahead and merge after review.

File: src/jit/lsrabuild.cpp [hunk 1]
```diff
@@ -2761,6 +2761,10 @@ int LinearScan::BuildOperandUses(GenTree* node, regMaskTP candidates)
     {
         return BuildIndirUses(node->AsIndir(), candidates);
     }
+    if (node->OperIs(GT_LEA))
+    {
+        BuildAddrUses(node, candidates);
```
Comment (mikedn) [inline src/jit/lsrabuild.cpp]: Shouldn't this be a `return`?

**Comment by echesakovMSFT** [comment]
@CarolEidt 
> I've retried the arm leg without success, so my inclination is to go ahead and merge after review.

The second attempt has actual failure during test running:

```
    JIT/SIMD/Vector3_r/Vector3_r.sh [FAIL]
      Inconsistency detected by ld.so: dl-tls.c: 481: _dl_allocate_tls_init: Assertion `listp->slotinfo[cnt].gen <= GL(dl_tls_generation)' failed!
      
      Return code:      1
      Raw output file:      /root/helix/work/workitem/JIT/SIMD/Reports/JIT.SIMD/Vector3_r/Vector3_r.output.txt
      Raw output:
      BEGIN EXECUTION
      /root/helix/work/correlation/corerun Vector3_r.exe ''
      Expected: 100
      Actual: 127
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/root/helix/work/correlation
      > /root/helix/work/workitem/JIT/SIMD/Vector3_r/Vector3_r.sh
      Expected: True
      Actual:   False
      Stack Trace:
        /__w/1/s/bin/tests/Linux.arm.Checked/TestWrappers/JIT.SIMD/JIT.SIMD.XUnitWrapper.cs(3696,0): at JIT_SIMD._Vector3_r_Vector3_r_._Vector3_r_Vector3_r_sh()
```

I doubt that this is caused by these changes - it happens in other PRs as well - I will open an issue

File: src/jit/lsrabuild.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/jit/lsrabuild.cpp]: Yes, it should - thanks!

**Comment by CarolEidt** [comment]
The arm leg failed with #21504 again - same assert but a different test: JIT/Methodical/casts/SEH/_dbgthrow/_dbgthrow.sh

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib ping

**Pull request merged**